### PR TITLE
[feat] Multi run ckpt

### PR DESCRIPTION
**Pull request by Jackmin801**
> [!NOTE]
> Introduces per-run checkpointing for multi-run RL and updates orchestration/training flows to save, resume, and clean checkpoints safely.
> 
> - New `MultiCheckpointManager` with per-run `RunState` to save adapter weights, optimizer, and scheduler; supports resume via latest or specified step
> - Trainer splits single-run vs multi-run paths: keeps existing full checkpoints for single-run; uses `MultiCheckpointManager.save()` and per-run resume callbacks for multi-run; scheduler/optimizer add ordered post-creation hooks
> - Orchestrator resumes from checkpoints and now derives weight directory via `get_weight_dir`, preferring `checkpoints/step_*/weight` then `broadcasts/step_*`
> - Runs system enhanced with deletion hooks and progress syncing across ranks; improves run lifecycle handling and cleanup
> - Filesystem weight broadcast adds debug logging and cleans up run dirs when deleted
> - CI configs add `[ckpt]` to orchestrator and extend model settings; integration test rewritten to start/kill/resume orchestrators, verify STABLE checkpoints, and assert reward behavior
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 6aa808958d56706a3abd293dc32e4665d2284a73. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

**Comment by Jackmin801** [comment]
@codex review

File: src/prime_rl/trainer/ckpt.py [hunk 1]
```diff
@@ -339,7 +342,7 @@ def save(
             self.save_to_path(step_path, state_dict, lora_state_dict, model, tokenizer)
             # Write STABLE file to indicate checkpoint is complete (for eval to safely read)
             (step_path / "STABLE").touch()
-            self.ckpt_steps.append(step)
+        self.ckpt_steps.append(step)
```
[the remark this answers is not in the document]
Comment (Jackmin801) [reply src/prime_rl/trainer/ckpt.py:345]: na bro, we exit for non master in the clean function. codex ngmi

File: src/prime_rl/trainer/rl/train.py [hunk 2]
```diff
@@ -170,6 +176,12 @@ def compute_scaling(orch_config) -> float:
         optimizer = setup_multi_optimizer(config.optim, parallel_dims)
         scheduler = setup_multi_scheduler(optimizer, config.scheduler, config.max_steps)
 
+        # Register checkpoint loading callback at index 1 (after scheduler creation at index 0)
+        def load_run_checkpoint(_optimizer, idx: int) -> None:
+            ckpt_manager.load_run(idx, optimizer, scheduler)
```
[the remark this answers is not in the document]
Comment (Jackmin801) [reply src/prime_rl/trainer/rl/train.py:181]: The closure is needed because we assume we get the multi object. Could be made nicer I guess but this is not a bug

File: src/prime_rl/trainer/multi_ckpt.py [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+        self.managers: dict[int, CheckpointManager] = {}
+        self.runs.register_deletion_hook(self._run_deleted_hook)
+        self.runs.register_creation_hook(self._run_created_hook)
+
+    def _run_deleted_hook(self, idx: int, run_id: str) -> None:
+        if idx in self.managers:
+            del self.managers[idx]
+
+    def _run_created_hook(self, idx: int, run_id: str) -> None:
+        self.managers[idx] = self._get_or_create_manager(idx)
```
[the remark this answers is not in the document]
Comment (Jackmin801) [reply src/prime_rl/trainer/multi_ckpt.py]: its a bit weird but it should be fine because its only none when the orch has none ckpt config and if it does, the manager never gets summoned

File: src/prime_rl/trainer/multi_ckpt.py [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+                if self.world.is_master:
+                    run_dir = self.runs.get_run_dir(idx)
+                    broadcast_src = run_dir / "broadcasts" / f"step_{step}"
+                    weight_dst = run_dir / "checkpoints" / f"step_{step}" / "weight"
+                    shutil.copytree(broadcast_src, weight_dst)
+            except FileNotFoundError:
+                self.logger.warning(f"Run {idx} deleted during checkpoint, skipping")
+            except Exception as e:
+                self.logger.error(f"Error checkpointing run {idx}: {e}")
+            dist.barrier()
```
[the remark this answers is not in the document]
Comment (Jackmin801) [reply src/prime_rl/trainer/multi_ckpt.py:157]: we have a bug if so and they should be stuck

File: src/prime_rl/trainer/multi_ckpt.py [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+                # Copy broadcast folder to checkpoint
+                if self.world.is_master:
+                    run_dir = self.runs.get_run_dir(idx)
+                    broadcast_src = run_dir / "broadcasts" / f"step_{step}"
+                    weight_dst = run_dir / "checkpoints" / f"step_{step}" / "weight"
+                    shutil.copytree(broadcast_src, weight_dst)
+            except FileNotFoundError:
+                self.logger.warning(f"Run {idx} deleted during checkpoint, skipping")
+            except Exception as e:
+                self.logger.error(f"Error checkpointing run {idx}: {e}")
```
[the remark this answers is not in the document]
Comment (Jackmin801) [reply src/prime_rl/trainer/multi_ckpt.py:156]: yea thats fine

File: configs/ci/integration/rl_multi_run/orchestrator.toml [hunk 6]
```diff
@@ -17,3 +17,6 @@ max_tokens = 128
 [[env]]
 id = "reverse-text"
 
+[ckpt]
+interval = 10
+resume_step = -1
```
[the remark this answers is not in the document]
Comment (Jackmin801) [reply configs/ci/integration/rl_multi_run/orchestrator.toml:22]: this is a test mate

File: tests/integration/test_rl_multi_run_lora.py [hunk 7]
```diff
@@ -193,39 +342,55 @@ def multi_run_result(
...
             lines = strip_escape_codes(f.read()).splitlines()
-        check_number_in_range(lines, step=7, min_threshold=0.2, max_threshold=0.6, pattern=r"Reward:\s*(\d+\.\d{4})")
-        check_number_in_range(lines, min_threshold=0.65, pattern=r"Reward:\s*(\d+\.\d{4})")
+        if name in ["beta", "gamma"]:
+            check_number_in_range(
+                lines, step=7, min_threshold=0.2, max_threshold=0.6, pattern=r"Reward:\s*(\d+\.\d{4})"
+            )
+            check_number_in_range(lines, min_threshold=0.65, pattern=r"Reward:\s*(\d+\.\d{4})")
+        elif name in ["alpha_resume", "beta_resume"]:
+            check_number_in_range(lines, min_threshold=0.65, pattern=r"Reward:\s*(\d+\.\d{4})")
```
[the remark this answers is not in the document]
Comment (Jackmin801) [reply tests/integration/test_rl_multi_run_lora.py:393]: wuut? does not go up because they are already above 0.65.

File: tests/integration/test_rl_multi_run_lora.py [hunk 8]
```diff
@@ -107,79 +157,178 @@ def multi_run_result(
...
+    # ---------------------------
+    # We cant use the same dir in case the trainer misses the change
+    run_dir = output_dir / "run_alpha_resume"
+    ckpt_dir = run_dir / "checkpoints" / "step_10"
+    ckpt_dir.parent.mkdir(parents=True, exist_ok=True)
+    shutil.copytree(tmp_path / "alpha_ckpt_step_10", ckpt_dir)
+    print(f"Copied alpha checkpoint to {ckpt_dir}")
+    orch_procs["alpha_resume"] = start_orchestrator(
+        "alpha_resume", max_steps=20, output_dir=output_dir, wandb_project=wandb_project, wandb_name=wandb_name
+    )
```
[the remark this answers is not in the document]
Comment (Jackmin801) [reply tests/integration/test_rl_multi_run_lora.py:265]: its a tmp_path brother. it will be different on reruns

File: tests/integration/test_rl_multi_run_lora.py [hunk 9]
```diff
@@ -107,79 +157,178 @@ def multi_run_result(
...
+        killed_log,
+        conditions=["Step 11", "Step 12", "Step 13"],
+        proc=orch_procs["alpha"],
+        sigterm=True,
+    )
+
+    # Wait for trainer checkpoints to be saved (STABLE file indicates checkpoint is complete)
+    alpha_ckpt_dir = output_dir / "run_alpha" / "checkpoints" / "step_10"
+    wait_for_file(alpha_ckpt_dir / "STABLE", timeout=TIMEOUT)
+
```
[the remark this answers is not in the document]
Comment (Jackmin801) [reply tests/integration/test_rl_multi_run_lora.py:245]: the trainers checkpoint save is independent of the orchestrator this is fine.

File: src/prime_rl/trainer/multi_ckpt.py [hunk 10]
```diff
@@ -0,0 +1,219 @@
...
+                raise FileNotFoundError(f"Checkpoint not found at {ckpt_path}")
+            self.logger.info(f"Loading checkpoint from {ckpt_path}")
+            state_dict = torch.load(ckpt_path / f"rank_{self.world.rank}.pt")
+            app_state.load_state_dict(state_dict)
+
+            self.logger.info(f"Resumed run {self.runs.idx_2_id[idx]} from step {step}")
+            return True
+        except Exception as e:
+            self.logger.error(f"Error loading checkpoint for run {idx}: {e}")
+            return False
```
[the remark this answers is not in the document]
Comment (Jackmin801) [reply src/prime_rl/trainer/multi_ckpt.py:200]: We dont want to crash the trainer over one run not being able to load its checkpoint. I think ideally we have a mechanism to kick runs the trainer doesnt like: e.g. tries to write empty samples, taking too long, has bad config. but lets do this in future PR

File: src/prime_rl/trainer/rl/train.py [hunk 11]
```diff
@@ -134,17 +135,22 @@ def compute_scaling(orch_config) -> float:
...
+        ckpt_manager, weight_ckpt_manager = setup_ckpt_managers(config.output_dir, config.ckpt, config.model.lora)
+
+        if config.ckpt and config.ckpt.resume_step is not None and ckpt_manager is not None:
+            if config.ckpt.resume_step == -1:
+                checkpoint_step = resolve_latest_ckpt_step(ckpt_manager.ckpt_dir)
+            else:
+                checkpoint_step = config.ckpt.resume_step
+    else:
+        # Multi-run uses per-run checkpointing via MultiCheckpointManager
+        ckpt_manager, weight_ckpt_manager = setup_multi_checkpoint_manager(config.output_dir)
```
Comment (samsja) [inline src/prime_rl/trainer/rl/train.py:152]: we don't handle resume_step -1 ?

File: src/prime_rl/trainer/multi_ckpt.py [hunk 12]
```diff
@@ -0,0 +1,211 @@
...
+        scheduler: "MultiLoRAScheduler",
+    ) -> None:
+        for idx in self.runs.used_idxs:
+            step = self.runs.progress[idx].step
+            if not self._should_save(idx, step):
+                continue
+
+            manager = self.managers[idx]
+
+            try:
```
Comment (samsja) [inline src/prime_rl/trainer/multi_ckpt.py:126]: would add comment that we do need this large try expect during live multi lora run

**Review by samsja — APPROVED** [review]
added minor comments, pre-approving tho, feel free to merge when ready

File: src/prime_rl/trainer/rl/train.py (hunk 11 above)
Comment (Jackmin801) [reply src/prime_rl/trainer/rl/train.py:152]: we do but its in MultiCheckpointManager.load_run which gets called at run_creation. the reason is we follow the ckpt config of the orchestrator. not the trainer. so the trainer doesnt set any ckpt config

File: src/prime_rl/trainer/multi_ckpt.py (hunk 12 above)
Comment (Jackmin801) [reply src/prime_rl/trainer/multi_ckpt.py:126]: added comment on having the really wide try except. this will probably be more obvious in the future when trainer has the ability to crash orchestrators that are misbehaving but atm we just move on
